File: renderdocui/Controls/PipelineFlowchart.cs
/******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015-2017 Baldur Karlsson
 * Copyright (c) 2014 Crytek
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/


using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace renderdocui.Controls
{
    public partial class PipelineFlowchart : UserControl
    {
        public PipelineFlowchart()
        {
            InitializeComponent();

            this.DoubleBuffered = true;
            SetStyle(ControlStyles.AllPaintingInWmPaint | ControlStyles.UserPaint | ControlStyles.OptimizedDoubleBuffer, true);
        }

        public void SetStages(KeyValuePair<string, string>[] stages)
        {
            m_StageNames = stages;
            m_StagesEnabled = new bool[stages.Length];
            m_StageFlows = new bool[stages.Length];
            for(int i=0; i < stages.Length; i++)
                m_StageFlows[i] = true;

            Invalidate();
        }

        #region Events

        private static readonly object SelectedStageChangedEvent = new object();

        [Description("Event raised when the selected pipeline stage is changed."), Category("Behavior")]
        public event EventHandler<EventArgs> SelectedStageChanged
        {
            add { Events.AddHandler(SelectedStageChangedEvent, value); }
            remove { Events.RemoveHandler(SelectedStageChangedEvent, value); }
        }
        protected virtual void OnSelectedStageChanged(EventArgs e)
        {
            EventHandler<EventArgs> handler = (EventHandler<EventArgs>)Events[SelectedStageChangedEvent];
            if (handler != null)
                handler(this, e);
        }

        #endregion

        #region Data Properties

        private bool[] m_StagesEnabled = null;
        private bool[] m_StageFlows = null;
        private KeyValuePair<string, string>[] m_StageNames = null;

        public void SetStagesEnabled(bool[] enabled)
        {
            if (m_StagesEnabled != null && enabled.Length == m_StagesEnabled.Length)
                m_StagesEnabled = enabled;

            Invalidate();
        }

        public void SetStageName(int index, KeyValuePair<string, string> name)
        {
            if (index >= 0 && index < m_StageNames.Length)
                m_StageNames[index] = name;

            Invalidate();
        }

        public void IsolateStage(int index)
        {
            if (m_StageNames != null && index >= 0 && index < m_StageNames.Length)
                m_StageFlows[index] = false;
        }

        private bool IsStageEnabled(int index)
        {
            return m_StagesEnabled != null && m_StagesEnabled[index];
        }

        private int m_HoverStage = -1;
        private int m_SelectedStage = 0;

        [Browsable(false)]
        public int SelectedStage
        {
            get
            {
                return m_SelectedStage;
            }
            set
            {
                if (value >= 0 && m_StageNames != null && value < m_StageNames.Length)
                {
                    m_SelectedStage = value;
                    Invalidate();

                    OnSelectedStageChanged(new EventArgs());
                }
            }
        }

        #endregion

        #region Constants and positions/dimensions

        const int BoxBorderWidth = 3;
        const int MinBoxDimension = 25;
        const int MaxBoxCornerRadius = 20;
        const float BoxCornerRadiusFraction = 1.0f / 6.0f;
        const float ArrowHeadSize = 6.0f;
        const int MinBoxMargin = 4;
        const int BoxLabelMargin = 8;
        const float BoxMarginFraction = 0.02f;

        private Rectangle TotalAreaRect
        {
            get
            {
                Rectangle rect = ClientRectangle;

                rect.Inflate(-1, -1);

                rect.X += Padding.Left;
                rect.Width -= Padding.Left + Padding.Right;

                rect.Y += Padding.Top;
                rect.Height -= Padding.Top + Padding.Bottom;

                rect.Inflate(-BoxBorderWidth, -BoxBorderWidth);

                return rect;
            }
        }

        private float BoxMargin
        {
            get
            {
                float margin = Math.Max(TotalAreaRect.Width, TotalAreaRect.Height) * BoxMarginFraction;

                margin = Math.Max(MinBoxMargin, margin);

                return margin;
            }
        }

        private int NumGaps { get { return m_StageNames == null ? 1 : m_StageNames.Length - 1; } }
        private int NumItems { get { return m_StageNames == null ? 2 : m_StageNames.Length; } }

        private SizeF BoxSize
        {
            get
            {
                float boxeswidth = TotalAreaRect.Width - NumGaps * BoxMargin;

                float boxdim = Math.Min(TotalAreaRect.Height, boxeswidth / NumItems);

                boxdim = Math.Max(MinBoxDimension, boxdim);

                float oblongwidth = Math.Max(0, (boxeswidth - boxdim * NumItems) / NumItems);

                return new SizeF(boxdim + oblongwidth, boxdim);
            }
        }

        private RectangleF GetBoxRect(int i)
        {
            return new RectangleF(TotalAreaRect.X + i * (BoxSize.Width + BoxMargin),
                                  TotalAreaRect.Y + TotalAreaRect.Height / 2 - BoxSize.Height / 2,
                                  BoxSize.Width, BoxSize.Height);
        }

        #endregion

        #region Painting

        GraphicsPath RoundedRect(RectangleF rect, int radius)
        {
            GraphicsPath ret = new GraphicsPath();
            
            ret.StartFigure();

            ret.AddArc(rect.X, rect.Y, 2 * radius, 2 * radius, 180, 90);
            ret.AddLine(rect.X + radius, rect.Y, rect.X + rect.Width - radius, rect.Y);
            ret.AddArc(rect.X + rect.Width - radius * 2, rect.Y, 2 * radius, 2 * radius, 270, 90);
            ret.AddLine(rect.X + rect.Width, rect.Y + radius, rect.X + rect.Width, rect.Y + rect.Height - radius);
            ret.AddArc(rect.X + rect.Width - 2*radius, rect.Y + rect.Height - 2 * radius, 2 * radius, 2 * radius, 0, 90);
            ret.AddLine(rect.X + rect.Width - radius, rect.Y + rect.Height, rect.X + radius, rect.Y + rect.Height);
            ret.AddArc(rect.X, rect.Y + rect.Height - 2*radius, 2 * radius, 2 * radius, 90, 90);

            ret.CloseFigure();

            return ret;
        }

        void DrawArrow(Graphics dc, Brush b, Pen p, float headsize, float y, float left, float right)
        {
            dc.DrawLine(p, new PointF(left, y), new PointF(right, y));

            using (GraphicsPath head = new GraphicsPath())
            {
                head.StartFigure();

                head.AddLine(new PointF(right, y), new PointF(right - headsize, y - headsize));
                head.AddLine(new PointF(right - headsize, y - headsize), new PointF(right - headsize, y + headsize));

                head.CloseFigure();

                dc.FillPath(b, head);
            }
        }

        private void PipelineFlowchart_Paint(object sender, PaintEventArgs e)
        {
            if (m_StageNames == null)
                return;

            var dc = e.Graphics;

            dc.CompositingQuality = CompositingQuality.HighQuality;
            dc.SmoothingMode = SmoothingMode.AntiAlias;

            int radius = (int)Math.Min(MaxBoxCornerRadius, BoxSize.Height * BoxCornerRadiusFraction);

            float arrowY = TotalAreaRect.Y + TotalAreaRect.Height / 2;

            using (var pen = new Pen(SystemBrushes.WindowFrame, BoxBorderWidth))
            using (var selectedpen = new Pen(Brushes.Red, BoxBorderWidth))
            {
                for (int i = 0; i < NumGaps; i++)
                {
                    if (!m_StageFlows[i] || !m_StageFlows[i + 1])
                        continue;

                    float right = TotalAreaRect.X + (i + 1) * (BoxSize.Width + BoxMargin);
                    float left = right - BoxMargin;

                    DrawArrow(dc, SystemBrushes.WindowFrame, pen, ArrowHeadSize, arrowY, left, right);
                }

                for (int i = 0; i < NumItems; i++)
                {
                    RectangleF boxrect = GetBoxRect(i);

                    var backBrush = SystemBrushes.Window;
                    var textBrush = SystemBrushes.WindowText;
                    var outlinePen = pen;

                    if (SystemInformation.HighContrast)
                    {
                        backBrush = SystemBrushes.ActiveCaption;
                        textBrush = SystemBrushes.ActiveCaptionText;
                    }

                    if (!IsStageEnabled(i))
                    {
                        backBrush = SystemBrushes.InactiveCaption;
                        textBrush = SystemBrushes.InactiveCaptionText;
                    }

                    if (i == m_HoverStage)
                    {
                        backBrush = SystemBrushes.Info;
                        textBrush = SystemBrushes.InfoText;
                    }

                    if (i == SelectedStage)
                    {
                        //backBrush = Brushes.Coral;
                        outlinePen = selectedpen;
                    }

                    using (var boxpath = RoundedRect(boxrect, radius))
                    {
                        dc.FillPath(backBrush, boxpath);
                        dc.DrawPath(outlinePen, boxpath);
                    }

                    StringFormat format = new StringFormat();
                    format.LineAlignment = StringAlignment.Center;
                    format.Alignment = StringAlignment.Center;

                    var s = m_StageNames[i].Value;

                    var size = dc.MeasureString(s, Font);

                    // Decide whether we can draw the whole stage name or just the abbreviation.
                    // This can look a little awkward sometimes if it's sometimes abbreviated sometimes
                    // not. Maybe it should always abbreviate or not at all?
                    if (size.Width + BoxLabelMargin > (float)boxrect.Width)
                    {
                        s = s.Replace(" ", "\n");
                        size = dc.MeasureString(s, Font);

                        if (size.Width + BoxLabelMargin > (float)boxrect.Width ||
                            size.Height + BoxLabelMargin > (float)boxrect.Height)
                        {
                            s = m_StageNames[i].Key;
                            size = dc.MeasureString(s, Font);
                        }
                    }

                    dc.DrawString(s, Font, textBrush, new PointF(boxrect.X + boxrect.Width / 2, arrowY), format);
                }
            }
        }

        #endregion

        #region Mouse Handling

        protected override void OnMouseMove(MouseEventArgs e)
        {
            base.OnMouseMove(e);

            int old = m_HoverStage;
            m_HoverStage = -1;

            for(int i=0; i < NumItems; i++)
            {
                if (GetBoxRect(i).Contains(e.Location))
                {
                    m_HoverStage = i;
                    break;
                }
            }

            if(m_HoverStage != old)
                Invalidate();
        }

        protected override void OnMouseLeave(EventArgs e)
        {
            base.OnMouseLeave(e);

            m_HoverStage = -1;
            Invalidate();
        }

        protected override void OnMouseClick(MouseEventArgs e)
        {
            base.OnMouseClick(e);

            for (int i = 0; i < NumItems; i++)
            {
                if (GetBoxRect(i).Contains(e.Location))
                {
                    SelectedStage = i;
                    break;
                }
            }
        }

        #endregion
    }
}

File: qrenderdoc/Widgets/PipelineFlowChart.cpp
/******************************************************************************
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2019 Baldur Karlsson
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#include "PipelineFlowChart.h"
#include <QMouseEvent>
#include <QPainter>

PipelineFlowChart::PipelineFlowChart(QWidget *parent) : QFrame(parent)
{
  setMouseTracking(true);
}

PipelineFlowChart::~PipelineFlowChart()
{
}

void PipelineFlowChart::setStages(const QStringList &abbrevs, const QStringList &names)
{
  m_StageNames = names;
  m_StageAbbrevs = abbrevs;
  m_StageFlows.reserve(m_StageNames.count());
  m_StagesEnabled.reserve(m_StageNames.count());
  for(int i = 0; i < m_StageNames.count(); i++)
  {
    m_StageFlows.push_back(true);
    m_StagesEnabled.push_back(true);
  }

  update();
}

void PipelineFlowChart::setStageName(int index, const QString &abbrev, const QString &name)
{
  if(index >= 0 && index < m_StageNames.count())
  {
    m_StageAbbrevs[index] = abbrev;
    m_StageNames[index] = name;
  }

  update();
}

void PipelineFlowChart::setIsolatedStage(int index)
{
  if(index >= 0 && index < m_StageFlows.count())
    m_StageFlows[index] = false;
}

void PipelineFlowChart::setSelectedStage(int idx)
{
  if(idx >= 0 && idx < m_StageNames.length())
  {
    m_SelectedStage = idx;
    update();

    emit stageSelected(idx);
  }
}

int PipelineFlowChart::selectedStage()
{
  return m_SelectedStage;
}

QSize PipelineFlowChart::minimumSizeHint() const
{
  return QSize(0, 30);
}

QSize PipelineFlowChart::sizeHint() const
{
  return QSize(0, 60);
}

void PipelineFlowChart::setStagesEnabled(const QList<bool> &enabled)
{
  for(int i = 0; i < enabled.count() && i < m_StagesEnabled.count(); i++)
    m_StagesEnabled[i] = enabled[i];

  update();
}

bool PipelineFlowChart::stageEnabled(int index)
{
  return index >= 0 && index < m_StagesEnabled.count() && m_StagesEnabled[index];
}

int PipelineFlowChart::numGaps()
{
  if(!m_StageNames.isEmpty())
    return m_StageNames.count() - 1;

  return 1;
}

int PipelineFlowChart::numItems()
{
  if(!m_StageNames.isEmpty())
    return m_StageNames.count();

  return 2;
}

QRectF PipelineFlowChart::totalAreaRect()
{
  QRectF rect = this->rect();

  qreal m = 6 + BoxBorderWidth;

  return rect.marginsRemoved(QMarginsF(m, m, m, m));
}

qreal PipelineFlowChart::boxMargin()
{
  qreal margin = qMax(totalAreaRect().width(), totalAreaRect().height()) * BoxMarginFraction;

  margin = qMax(MinBoxMargin, margin);

  return margin;
}

QRectF PipelineFlowChart::boxRect(int i)
{
  QRectF totalRect = totalAreaRect();

  qreal boxeswidth = totalRect.width() - numGaps() * boxMargin();

  qreal boxdim = qMin(totalRect.height(), boxeswidth / numItems());

  boxdim = qMax(MinBoxDimension, boxdim);

  qreal oblongwidth = qMax(0.0, (boxeswidth - boxdim * numItems()) / numItems());

  QSizeF boxSize(boxdim + oblongwidth, boxdim);

  QRectF ret(totalRect.x() + i * (boxSize.width() + boxMargin()),
             totalRect.y() + totalRect.height() / 2 - boxSize.height() / 2, boxSize.width(),
             boxSize.height());

  return ret;
}

void PipelineFlowChart::drawArrow(QPainter &p, const QPen &pen, float headsize, float y, float left,
                                  float right)
{
  p.setPen(pen);

  p.drawPolygon(QPolygonF({QPointF(right, y), QPointF(right - headsize, y - headsize),
                           QPointF(right - headsize, y + headsize)}));

  QPen linePen(pen);
  linePen.setWidthF(BoxBorderWidth);
  p.setPen(linePen);

  p.drawLine(QPointF(left, y), QPointF(right, y));
}

void PipelineFlowChart::paintEvent(QPaintEvent *e)
{
  if(m_StageNames.empty())
    return;

  QPainter p(this);

  p.fillRect(rect(), Qt::transparent);

  p.setRenderHint(QPainter::Antialiasing, true);

  const QRectF totalRect = totalAreaRect();
  const QRectF box0Rect = boxRect(0);

  const qreal radius = qMin(MaxBoxCornerRadius, box0Rect.height() * BoxCornerRadiusFraction);

  const qreal arrowY = totalRect.y() + totalRect.height() / 2;

  QColor enabledBase = palette().color(QPalette::Base);
  QColor enabledText = palette().color(QPalette::Text);
  QColor disabledBase = palette().color(QPalette::Disabled, QPalette::Base);
  QColor disabledText = palette().color(QPalette::Disabled, QPalette::Text);
  QColor tooltip = palette().color(QPalette::ToolTipBase);
  QColor tooltipText = palette().color(QPalette::ToolTipText);

  QPen pen(enabledText);
  QPen selectedPen(Qt::red);

  int num = numGaps();
  for(int i = 0; i < num; i++)
  {
    if(!m_StageFlows[i] || !m_StageFlows[i + 1])
      continue;

    float right = totalRect.x() + (i + 1) * (box0Rect.width() + boxMargin());
    float left = right - boxMargin();

    p.setBrush(enabledText);
    drawArrow(p, pen, ArrowHeadSize, arrowY, left, right);
  }

  num = numItems();
  for(int i = 0; i < num; i++)
  {
    QRectF boxrect = boxRect(i);

    QBrush backBrush(enabledBase);
    QPen textPen(enabledText);
    QPen outlinePen(enabledText);

    if(!stageEnabled(i))
    {
      backBrush.setColor(disabledBase);
      outlinePen.setColor(disabledText);
    }

    if(i == m_HoverStage)
    {
      backBrush.setColor(tooltip);
      textPen.setColor(tooltipText);
    }

    if(i == m_SelectedStage)
    {
      outlinePen = selectedPen;
    }

    outlinePen.setWidthF(BoxBorderWidth);

    p.setPen(outlinePen);
    p.setBrush(backBrush);
    p.drawRoundedRect(boxrect, radius, radius);

    QTextOption opts(Qt::AlignCenter);
    opts.setWrapMode(QTextOption::NoWrap);

    QString s = m_StageNames[i];

    QRectF reqBox = p.boundingRect(QRectF(0, 0, 1, 1), m_StageNames[i], opts);

    if(reqBox.width() + BoxLabelMargin > (float)boxrect.width())
      s = m_StageAbbrevs[i];

    p.setPen(textPen);
    p.drawText(boxrect, s, opts);
  }
}

void PipelineFlowChart::mouseMoveEvent(QMouseEvent *e)
{
  int old = m_HoverStage;
  m_HoverStage = -1;

  const int num = numItems();
  for(int i = 0; i < num; i++)
  {
    if(boxRect(i).contains(e->pos()))
    {
      m_HoverStage = i;
      break;
    }
  }

  if(m_HoverStage != old)
    update();
}

void PipelineFlowChart::mousePressEvent(QMouseEvent *e)
{
  if(e->button() == Qt::LeftButton)
  {
    const int num = numItems();
    for(int i = 0; i < numItems(); i++)
    {
      if(boxRect(i).contains(e->pos()))
      {
        setSelectedStage(i);
        break;
      }
    }
  }
}

void PipelineFlowChart::leaveEvent(QEvent *e)
{
  m_HoverStage = -1;

  update();
}

void PipelineFlowChart::focusOutEvent(QFocusEvent *event)
{
  m_HoverStage = -1;

  update();
}
